cprover_bindings/src/irep/goto_binary_serde.rs

```diff
@@ -0,0 +1,1462 @@
+///
+/// A [NumberedIrep] is conceptually a pair made of the [IrepKey] itself and the
+/// unique number assigned to that key.
+///
+/// The cache implemented by [IrepNumbering] is bidirectional. It lets you
+/// compute the [NumberedIrep] of an [Irep], and lets you fetch a numbered
+/// [NumberedIrep] from its unique number.
+///
+/// In practice:
+/// - the forward directon from [IrepKey] to unique numbers is
+/// implemented using a `HashMap<IrepKey,usize>`
+/// - the inverse direction from unique numbers to [NumberedIrep] is implemented
+/// using a `Vec<NumberedIrep>` called the `index` that stores [NumberedIrep]
+/// under their unique number.
+///
+/// Earlier we said that an [NumberedIrep] is conceptually a pair formed of
+/// an [IrepKey] and its unique number. In practice, is  represented using only
+/// a pair formed of a `usize` representing the unique number, and a `usize`
+/// representing the index at which the key can be found inside a single vector
+/// of type `Vec<usize>` called `keys` where all [IrepKey] are concatenated when
+/// they first get numbered. The inverse map of keys is represented this way
+/// because the Rust hash map that implements the forward cache owns
+/// its keys which would make it difficult to store keys references in inverse
+/// cache, which would introduce circular dependencies and require `Rc` and
+/// liftetimes annotations everywhere.
+///
+/// Numberig an [Irep] consists in recursively traversing it and numbering its
+/// contents in a bottom-up fashion, then assembling its [IrepKey] and querying
+/// the cache to either return an existing [NumberedIrep] or creating a new one
+/// (in case that key has never been seen before).
+///
+/// The [GotoBinarySerializer] internally uses a [IrepNumbering] and a cache
+/// of [NumberedIrep] and [NumberedString] it has already written to file.
+///
+/// When given an [InternedString], [IrepId] or [Irep] to serialize,
+/// the [GotoBinarySerializer] first numbers that object using its internal
+/// [IrepNumbering] instance. Then it looks up that unique number in its cache
+/// of already written objects. If the object was seen before, only the unique
+/// number of that object is written to file. If the object was never seen
+/// before, then the unique number of that object is written to file, followed
+/// by the objects describing its contents (themselves only being written fully
+/// if they have never been seen before, or only referenced if they have been
+/// seen before, etc.)
+///
+/// The [GotoBinaryDeserializer] also uses an [IrepNumbering] and a cache
+/// of [NumberedIrep] and [NumberedString] it has already read from file.
+/// Dually to the serializer, it will only attempt to decode the contents of an
+/// object from the byte stream on the first occurrence.
+
+/// A numbered [InternedString]. The number is guaranteed to be in [0,N].
+/// Had to introduce this indirection because [InternedString] does not let you
```

Comment:
I just found out that `InternedString` can be updated to allow getting the underlying usize value:
```diff
$ git diff
diff --git a/cprover_bindings/src/cbmc_string.rs b/cprover_bindings/src/cbmc_string.rs
index b487b761512..46956aebcf9 100644
--- a/cprover_bindings/src/cbmc_string.rs
+++ b/cprover_bindings/src/cbmc_string.rs
@@ -4,7 +4,7 @@
 use lazy_static::lazy_static;
 use std::sync::Mutex;
 use string_interner::symbol::SymbolU32;
-use string_interner::StringInterner;
+use string_interner::{StringInterner, Symbol};
 
 /// This class implements an interner for Strings.
 /// CBMC objects to have a large number of strings which refer to names: symbols, files, etc.
@@ -46,6 +46,10 @@ pub fn map<T, F: FnOnce(&str) -> T>(&self, f: F) -> T {
     pub fn starts_with(&self, pattern: &str) -> bool {
         self.map(|s| s.starts_with(pattern))
     }
+
+    pub fn to_usize(&self) -> usize {
+        self.0.to_usize()
+    }
 }
 
 impl std::fmt::Display for InternedString {
@@ -112,6 +116,9 @@ fn test_string_interner() {
         let a: InternedString = "A".into();
         let b: InternedString = "B".into();
         let aa: InternedString = "A".into();
+        println!("a.to_usize(): {}", a.to_usize());
+        println!("b.to_usize(): {}", b.to_usize());
+        println!("aa.to_usize(): {}", aa.to_usize());
 
         assert_eq!(a, aa);
         assert_ne!(a, b);
```
Running the unit test with `cargo test` yields:
```
$ cargo test -p cprover_bindings string_interner -- --nocapture
    Blocking waiting for file lock on build directory
   Compiling cprover_bindings v0.21.0 (/home/ubuntu/git/kani/cprover_bindings)
    Finished test [unoptimized + debuginfo] target(s) in 1.75s
     Running unittests src/lib.rs (target/debug/deps/cprover_bindings-96ecbe9fe7cb293d)

running 1 test
a.to_usize(): 0
b.to_usize(): 1
aa.to_usize(): 0
test cbmc_string::tests::test_string_interner ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 19 filtered out; finished in 0.00s
```
Would this be useful to simplify the structs?

Reply:
Hi, I tried that suggestion and it does not seem to allow to simplify things that much.

We still need to distinguish the concept of `InternedString` from that of `NumberedString`,  the key difference being that we know a `NumberedString` has been produced by the `IrepNumbering` and hence that its number is a valid index into the inverse cache of strings. 

Doing things like this also ensures our numbering is dense in [0,N] and independent of how `InternedString` does its own numbering, which allows us to use `Vec<NumberedString>` for the inverse cache. 

Using `InternedString` directly instead, the inverse cache turns into a vector of `Vec<Option<InternedString>>` or `HashMap<usize, InternedString>` and every lookup into the inverse cache requires extra checks.